### ISSUE TITLE
Sequence-Wise Batch-Norm correct?

### ISSUE THREAD

**Issue by ATriantafyllopoulos** [SeanNaren/deepspeech.pytorch#360]
I was looking at the code for sequence-wise batch-norm: https://github.com/SeanNaren/deepspeech.pytorch/blob/413e17626f835cad9082a6e029461fb202886561/model.py#L28

What this does is transform the input from shape `(T, N, H)` to `((TxN), H)`, and apply `BatchNorm1d` to it.

This will result in the formula:
![codecogseqn](https://user-images.githubusercontent.com/14807371/50901768-70c1d580-1419-11e9-983e-1c178b74e2d2.gif)

Instead of the one mentioned in https://arxiv.org/abs/1510.01378 (with formula slightly adapted for PyTorch syntax):
![codecogseqn2](https://user-images.githubusercontent.com/14807371/50901910-c72f1400-1419-11e9-9cee-217717d4c9c3.gif)

where n is the number of *unpadded frames* and `T` in this case being the sequence length of *each sequence*.

In my opinion, the two formulas differ in that the one in the current repo will take into account all the padded inputs for the computation of the mean (and the std accordingly), which is what sequence-wise batch-norm wants to *avoid*. It will only work in the case where all the sequences in the batch had the same length to begin with.

Is this something you took care of in the code (i.e. that all sequences in a mini-batch have the same length to begin with) or have I missed something here?

**Comment by ryanleary** [comment]
I believe you are correct that this is a bug. The good news is that the sequence lengths among a particular mini-batch tend to be close (though not exact), so there is some padding included in the batchnorm.

This is a tricky thing to fix, though I believe if you hack on the packed representation, you can work around it effectively, see https://github.com/ryanleary/patter/blob/92e8fb0d5d98571bd76a0ff90fae1fc1853ba0da/patter/layers/rnns.py#L24

**Comment by ATriantafyllopoulos** [comment]
Thanks a lot! It seems to me that your fix works properly. If you do not want to change anything in this package then feel free to close the issue.
